File: medium/2109_adding_spaces_to_a_string/solution.cpp
#include <string>
#include <vector>

class Solution {
public:
  std::string addSpaces(std::string s, std::vector<int> &spaces) {
    std::string sWithSpaces(s.size() + spaces.size(), ' ');

    for (size_t sReadIdx{}, sWriteIdx{}, spacesIdx{}; sReadIdx < s.size();
         ++sReadIdx, ++sWriteIdx) {
      if (spacesIdx < spaces.size() && sReadIdx == spaces[spacesIdx]) {
        ++sWriteIdx;
        ++spacesIdx;
      }

      sWithSpaces[sWriteIdx] = s[sReadIdx];
    }

    return sWithSpaces;
  }
};

File: medium/2109_adding_spaces_to_a_string/solution.py
class Solution:
    def addSpaces(self, s: str, spaces: list[int]) -> str:
        words = []

        prevSpace = 0
        for space in spaces:
            words.append(s[prevSpace:space])
            prevSpace = space

        words.append(s[prevSpace:])

        return " ".join(words)
